### DISCUSSION TITLE
Fix a 3D Model in 3D space

### DISCUSSION THREAD

**Discussion by inveders** [google/filament#3985] [Q&A] [answered]
I have a 3D model in .glb wich I see through my camera, and I want to keep it in the center of my screen, when I move my phone, and after a click on a floating button, I want to fix this model in the 3D space (360°) of the room I am in. But after clicking, the 3D model continue to move with my phone, in the center. How to fix it in the space?

This is my method to retrieve screen position (width and height are always the same, 400 and 800) and transform them in 3D coordinates (with V3 method)

 ```
private suspend fun initTranslation() {
        coroutineScope.launch {
            // translation
            firstPostModelEvents
                .mapNotNull { firstModelEvent ->
                    (firstModelEvent as? FirstPostModelEvent.Move)
                        ?.let {
                            arCore.frame
                                .hitTest(
                                    firstModelEvent.screenPosition.x,
                                    firstModelEvent.screenPosition.y,
                                )
                                .maxByOrNull { it.trackable is Point }
                        }
                        ?.let { V3(it.hitPose.translation) }
                }
                .collect {
                    canDrawBehavior.tryEmit(Unit)
                    translation = it
                }
        }
    }
 ```

and this is the method to "fix" my model in the 3D space

 ```
private fun setFirstTransform() {
        //permit to place the 3D model on the screen
        filamentAsset?.let {
            filament.engine?.transformManager?.let { transformManager ->
                transformManager.setTransform(
                    transformManager.getInstance(it.root),
                    m4Identity()
                        .translate(translation.x, translation.y, translation.z)
                        .rotate(rotate.toDegrees, 0f, 1f, 0f)
                        .scale(scale, scale, scale)
                        .floatArray,
                )
            }
        }
    }
 ```

What doesn't work? Why the model don't fix on the 3D space?

**Comment by romainguy** [comment, marked as answer]
I'm not extremely familiar with ARCore but if I remember correctly the positions of the objects in the world are recomputed on every frame. You need to create an [anchor](https://developers.google.com/ar/develop/developer-guides/anchors) which will give you a transform every frame. Then you either apply the anchor's transform to the object you want to keep in a fixed position in the world, or you create an invisible `Entity` and make it the parent your 3D object (and you'll need to update that invisible `Entity`'s transform every time ARCore gives you an updated transform for the anchor).

**Reply by inveders** [reply]
So the setTransform method is just to place the object on my screen and not in the 3D space? Because I already put in the position in the world. Or how to "apply the anchor's transform to the object"? There is an other way that with setTransform?

**Reply by romainguy** [reply]
No `setTransform()` applies a transform to move your object to its final position/orientation in world (3D) space. The final screenspace position is determined by the camera. You can combine multiple transforms by multiplying the matrices (it's done automatically when you use parents, so you could create an `Entity` whose transform comes from ARCore's anchor, and set an addition position/transform on your object as long as that object is a child of the Entity I just described). Hope it makes sense.

**Reply by inveders** [reply]
Ok I was thinking that in the initTranslation() method my 
```
arCore.frame.hitTest(firstModelEvent.screenPosition.x, firstModelEvent.screenPosition.y,)
```
was my anchor for ArCore, with the hitTest, but I think that it's not the case...
I understand the usefulness of setTransform(), but I don't understand all your explanation, especially with parents and child but I'll try to dig into the subject.

I have created an anchor : val anchor:Anchor = HitResult.createAnchor()
And I create a new invisible entity : val anchorEntity = EntityManager.get().create()
I add it on the scene : filament.scene?.addEntity(anchorEntity)
So my model become the child of the anchorEntity by this way : transformManager.setParent(modelEntity, anchorEntity)

But how to link my anchor with the anchorEntity? is by an other transformManager getting anchorEntity as instance and with the coordinates of the anchor?

**Reply by romainguy** [reply]
>But how to link my anchor with the anchorEntity? is by an other transformManager getting anchorEntity as instance and with the coordinates of the anchor?

ARCore's `Anchor` has a `getPose()` method which gives you the transform matrix (`toMatrix()`) that you can set on your `anchorEntity` using Filament's `TransformManager`.

**Reply by inveders** [reply]
Thanks you !

**Reply by ThomasGorisse** [reply]
Depending on your usage (more accurate from real world or more fixed position from the camera movements), you can also update your object position every frame using the `frame.getCamera().getDisplayOrientedPose()`.

It's not the ARCore recommended way but you won't have the object little moves because of improving his position every frame.

**Reply by yashukla47** [reply]
Hey @inveders, are you writing your own code to bind Filament with Arcore? how is it going?
